src/atom.rs

```diff
@@ -470,19 +470,21 @@ impl<Static: StaticAtomSet> Deserialize for Atom<Static> {
 // over the one from &str.
 impl<Static: StaticAtomSet> Atom<Static> {
     pub fn to_ascii_uppercase(&self) -> Self {
-        if self.chars().all(char::is_uppercase) {
-            self.clone()
-        } else {
-            Atom::from(&*((&**self).to_ascii_uppercase()))
+        for b in self.bytes() {
+            if let b'a' ... b'z' = b {
+                return Atom::from((&**self).to_ascii_uppercase())
```

Comment:
Is this the fast path you're mentioning? This here says that if any one of the byte characters in the atom is a lowercase, then we uppercase the entire atom.